### ISSUE TITLE
SingleInstance registration causes service to be instantiated multiple times when using nested lifetime scopes with the same tag

### ISSUE THREAD

**Issue by oliverhanappi** [autofac/Autofac#882]
When a service is registered as SingleInstance in a lifetime scope, all nested lifetime scopes should retrieve the instance owned by that parent. However, if a nested lifetime scope is created with the same tag as its parent, one instance per lifetime scope is created.

In the following example, the class MyService is instantiated twice. Once when resolving from the parent lifetime scope, once when resolving from the child lifetime scope. If no tags or different tags are used for the two lifetime scopes, it works as expected.

````csharp
using System;
using Autofac;

namespace AutofacPlayground
{
  public static class Program
  {
    public static void Main ()
    {
      var builder = new ContainerBuilder();
      using (var container = builder.Build())
      {
        using (var parentLifetimeScope = container.BeginLifetimeScope ("MyTag",
          b => b.RegisterType<MyService>().SingleInstance()))
        {
          var parentService = parentLifetimeScope.Resolve<MyService>();
          using (var childLifetimeScope = parentLifetimeScope.BeginLifetimeScope ("MyTag"))
          {
            var childService = childLifetimeScope.Resolve<MyService>();
            if (parentService != childService)
              Console.WriteLine ("Parent and Child are not the same.");
          }
        }
      }
    }
  
    public class MyService { }
  }
}
````

Tested using Autofac 4.6.2 on .NET 4.6.1.

**Comment by alexmg** [comment]
The problem here is that you have been explicit about what tagged lifetime scope to look for the service in when it was registered during the first `BeginLifetimeScope` call. When the resolve operation is performed in the second lifetime scope, which has been given the same tag as the first, the scope matching does not walk further up the hierarchy because it has already found a match. Because the second lifetime scope is considered a match, and doesn’t yet contain an activated instance, a new instance is created and returned.

You should avoid using duplicate tag names as they are meant to represent a hierarchy where searching continues upward towards the root lifetime scope until a match is found. I have added a check that ensures the tag provided to `BeginLifetimeScope` has not already been assigned to a parent lifetime scope.

**Comment by oliverhanappi** [comment]
Thanks for the clarification. It was indeed an accidential error that the same tag got assigned to a parent and child lifetime scope, however I was still surprised that the instance has been created multiple times. Glad to see that there is now a check which prevents such situations.

**Comment by tillig** [comment]
Fix released in [v4.7.0](https://www.nuget.org/packages/Autofac/4.7.0). (Exception is now thrown if nested scopes have the same tag as a parent.)
